### PR TITLE
[TF-TRT] Adding Graph Visualization Export Instrumentation

### PR DESCRIPTION
**Pull request by DEKHTIARJonathan**
This PR introduces a  TF-TRT graph visualization export tool in `graphviz` format.

The user can then convert the `graphviz` file in `.dot` format to a more commonly readable format like `PNG`.

Usage: `TF_TRT_EXPORT_GRAPH_VIZ_PATH=/path/to/mymodel.dot python main.py`

### CODE REVIEW

**Comment by DEKHTIARJonathan** [comment]
@bixia1 for review
CC: @tfeher @christopherbate

File: tensorflow/python/compiler/tensorrt/trt_convert.py [hunk 1]
```diff
@@ -1262,6 +1252,19 @@ def _save_calibration_table(node):
...
+    if graphviz_path is not None:
+      try:
+        trt_utils.export_graph_to_dot(
+          graphdef=self._converted_func.graph.as_graph_def(add_shapes=True),
+          dot_output_filename=graphviz_path
+        )
+      except Exception as e:
+        logging.error(
+            "An Exception occured during the export of the graph "
+            f"visualization: {str(e)}")
```
Comment (bixia1) [inline tensorflow/python/compiler/tensorrt/trt_convert.py:1266]: I think "e" is enough to print the string of e, no need to use "str(e)".

File: tensorflow/python/compiler/tensorrt/utils.py [hunk 2]
```diff
@@ -86,3 +88,97 @@ def is_experimental_feature_activated(feature_name):
   return (feature_name
           in os.environ.get("TF_TRT_EXPERIMENTAL_FEATURES",
                             default="").split(","))
+
+
+def get_nodes_in_engine(graphdef, node_name):
```
Comment (bixia1) [inline tensorflow/python/compiler/tensorrt/utils.py:93]: Can you document what this function does?

File: tensorflow/python/compiler/tensorrt/utils.py [hunk 3]
```diff
@@ -86,3 +88,97 @@ def is_experimental_feature_activated(feature_name):
...
+  for func in graphdef.library.function:
+    if f"{node_name}_native_segment" == func.signature.name:
+      node_count = len(func.node_def)
+      for node in func.node_def:
+        ops_in_engine[node.op] += 1
+      break
+  return node_count, ops_in_engine
+
+
+def export_graph_to_dot(graphdef, dot_output_filename):
```
Comment (bixia1) [inline tensorflow/python/compiler/tensorrt/utils.py:104]: Similar to the above, please add document to the function.

File: tensorflow/python/compiler/tensorrt/utils.py [hunk 4]
```diff
@@ -86,3 +88,97 @@ def is_experimental_feature_activated(feature_name):
...
+    print("    node [width=1.35];", file=f)
+    nodes_with_no_inputs = []
+    for node in graphdef.node:
+      output_name = node.name
+
+      try:
+        device_key = node.device.split("/")[-1]
+        device_clr_id = device_clr[device_key if device_key else "device:None"]
+      except:
+        device_clr_id = "white"
```
Comment (bixia1) [inline tensorflow/python/compiler/tensorrt/utils.py:145]: why do we need to "except" path?

File: tensorflow/python/compiler/tensorrt/utils.py [hunk 5]
```diff
@@ -86,3 +88,97 @@ def is_experimental_feature_activated(feature_name):
...
+
+      if node.op == "TRTEngineOp":
+        node_count, _ = get_nodes_in_engine(graphdef, output_name)
+        node_label = f"{output_name} [{node_count}]"
+        print(
+            f"  \"{output_name}\" [label=\"{output_name} [{node_count}]\"];",
+            file=f
+        )
```
Comment (bixia1) [inline tensorflow/python/compiler/tensorrt/utils.py:153]: Shall we allow user to point to the TRTEngine node to inspect the details there, in a follow PR?

**Comment by bixia1** [comment]
@DEKHTIARJonathan It is nice to show an example of the graph here. But would you please move the graph example out of the PR description and put it in a comment block?

File: tensorflow/python/compiler/tensorrt/utils.py (hunk 5 above)
Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/utils.py:153]: Yes there are multiple possible improvements we could bring to this graph feature. This is one of them. As for now it's a first iteration of this tool. I expect this tool to grow as we need.

**Comment by DEKHTIARJonathan** [comment]
Example with `Electra TF2`:

image

**Comment by gbaned** [comment]
@DEKHTIARJonathan Can you please resolve conflicts? Thanks!

**Comment by DEKHTIARJonathan** [comment]
@bixia1 I have finished reworking this PR. It now includes the following:
- compute device per node
- coloring per compute dtype

File: tensorflow/python/compiler/tensorrt/utils.py (hunk 4 above)
Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/utils.py]: That part was removed

File: tensorflow/python/compiler/tensorrt/trt_convert.py (hunk 1 above)
Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/trt_convert.py:1281]: Fixed

File: tensorflow/python/compiler/tensorrt/utils.py (hunk 3 above)
Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/utils.py]: Fixed

**Comment by DEKHTIARJonathan** [comment]
@bixia1 I have finished reworking this PR. It now includes the following:
- compute device per node
- coloring per compute dtype

Please review

File: tensorflow/python/compiler/tensorrt/utils.py [hunk 6]
```diff
@@ -86,3 +89,168 @@ def is_experimental_feature_activated(feature_name):
...
+        else:
+          return "unknown"
+      else:
+        return _convert_dtype_id_to_str(precision_val.type)
+    except Exception as e:
+      continue
+
+
+def get_node_io_shapes(node, key):
+  """Return the input/output shapes of a GraphDef Node."""
```
Comment (bixia1) [inline tensorflow/python/compiler/tensorrt/utils.py:132]: s/Return/Returns/

File: tensorflow/python/compiler/tensorrt/utils.py [hunk 7]
```diff
@@ -86,3 +89,168 @@ def is_experimental_feature_activated(feature_name):
...
+  return out_shape
+
+
+def get_trtengineop_io_dtypes(node, key):
+  """Return the input/output dtypes of a TRTEngineOp."""
+  return _convert_dtype_id_to_str(node.attr[key].list.type)
+
+
+def get_trtengineop_io_nodes_count(node, key):
+  """Return the number of input/output nodes of a TRTEngineOp."""
```
Comment (bixia1) [inline tensorflow/python/compiler/tensorrt/utils.py:145]: s/Return/Returns/

File: tensorflow/python/compiler/tensorrt/utils.py [hunk 8]
```diff
@@ -86,3 +89,168 @@ def is_experimental_feature_activated(feature_name):
...
+  return len(node.attr[key].list.type)
+
+
+def get_trtengineop_node_op_count(graphdef, node_name):
+  """Count the number of nodes and OP types of a given TRTEngineOp."""
```
Comment (bixia1) [inline tensorflow/python/compiler/tensorrt/utils.py:150]: s/Count/Counts/

File: tensorflow/python/compiler/tensorrt/utils.py [hunk 9]
```diff
@@ -86,3 +89,168 @@ def is_experimental_feature_activated(feature_name):
...
+def _convert_dtype_id_to_str(dtype):
+  """Helper function to convert a dtype id to a corresponding string name."""
+  if isinstance(dtype, int):
+    return dtypes._TYPE_TO_STRING[dtype]
+  else:
+    return [dtypes._TYPE_TO_STRING[d] for d in dtype]
+
+
+def get_node_compute_dtype(node):
+  """Return the compute DType of a GraphDef Node."""
```
[the remark this answers is not in the document]
Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/utils.py:103]: Fixed

File: tensorflow/python/compiler/tensorrt/utils.py [hunk 10]
```diff
@@ -86,3 +89,168 @@ def is_experimental_feature_activated(feature_name):
...
+  # Note: Order is important, please do not change.
```
[the remark this answers is not in the document]
Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/utils.py:104]: Fixed

File: tensorflow/python/compiler/tensorrt/utils.py [hunk 11]
```diff
@@ -86,3 +89,168 @@ def is_experimental_feature_activated(feature_name):
...
+def get_node_io_shapes(node, key):
+  """Return the input/output shapes of a GraphDef Node."""
+  out_shape = []
+  for shape in node.attr[key].list.shape:
+    out_shape.append([dim.size for dim in shape.dim])
+  return out_shape
+
+
+def get_trtengineop_io_dtypes(node, key):
+  """Return the input/output dtypes of a TRTEngineOp."""
```
[the remark this answers is not in the document]
Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/utils.py:140]: Fixed

File: tensorflow/python/compiler/tensorrt/utils.py (hunk 6 above)
Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/utils.py:132]: Fixed

File: tensorflow/python/compiler/tensorrt/utils.py (hunk 8 above)
Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/utils.py:150]: Fixed

File: tensorflow/python/compiler/tensorrt/utils.py (hunk 7 above)
Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/utils.py:145]: Fixed

**Comment by DEKHTIARJonathan** [comment]
@bixia1 comments addressed. Please review

**Review by bixia1 — APPROVED** [review]

**Comment by DEKHTIARJonathan** [comment]
@bixia1 any problem with this PR ? It's approved and not merged for 2 days

**Pull request merged**